Add Support for NeMo Customizer to Finetuning Harness

dnandakumar-nv opened the pull request:
## Description
This PR adds support for running Direct Preference Optimization (DPO) based finetuning using NeMo Customizer. 

Add's a NeMo Customizer package, finetuning components, an example, and comprehensive tests and docs. 

Depends on #1221. 


## By Submitting this PR I confirm:
- I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/NeMo-Agent-Toolkit/blob/develop/docs/source/resources/contributing.md).
- We require that all contributors "sign-off" on their commits. This certifies that the contribution is your original work, or you have rights to submit it under the same license, or a compatible license.
  - Any contribution which contains commits that are not Signed-Off will not be accepted.
- When the PR is ready for review, new or existing tests cover these changes.
- When the PR is ready for review, the documentation is up to date with these changes.


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Added DPO with NeMo Customizer (trajectory builder, trainer/adapter, evaluators, deployment) and a full Tic‑Tac‑Toe finetuning example with LLM/random move modes, multi-candidate pipelines, scoring, and selection.

* **Documentation**
  * New DPO guide, example README, and docs integrated into finetuning guides.

* **Tests**
  * Extensive unit and integration tests for DPO components, trainer adapter, and workflows.

* **Chores**
  * Updated CI path checks and allowlist patterns; added accepted vocabulary tokens and a small PR template link fix.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

dnandakumar-nv committed:
Update `uv.lock` to revise dependencies and add overrides

This commit updates the `uv.lock` file to increase the revision number and introduce new package overrides for `litellm` and `openai`. Additionally, metadata like upload times for files has been added, providing a more detailed package resolution record.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Remove OpenPipe ART finetuning backend implementation

Deleted the entire OpenPipe ART finetuning backend, including configurations, runner, and registration modules. This removal simplifies the codebase by eliminating components related to the ART backend that are no longer required.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Remove OpenPipe ART finetuning backend implementation

Deleted the entire OpenPipe ART finetuning backend, including configurations, runner, and registration modules. This removal simplifies the codebase by eliminating components related to the ART backend that are no longer required.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Add fine-tuning support with configurable trainer components

This commit introduces a `FinetuningConfig` model and integrates fine-tuning support across the codebase. It includes new components such as `Trainer`, `TrainerAdapter`, and `TrajectoryBuilder`, along with methods to manage and bind these components. Fine-tuning is now fully configurable and enabled through the updated workflow builder and configuration interfaces.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Rename FinetuningRunner to Trainer across the codebase

Replaced all instances of FinetuningRunner with Trainer to reflect updated naming conventions. This change ensures consistency in naming and improves clarity in describing the finetuning workflow orchestration. Updated relevant imports, class definitions, method signatures, and type hints.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Add fine-tuning trainer registration and type handling support

Introduced structures and methods to register, retrieve, and handle fine-tuning trainers, trainer adapters, and trajectory builders. Updated type registry and workflows to support these components, ensuring seamless integration into the fine-tuning process.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Refactor training components into distinct groups

Separated `TRAINING` into `TRAINERS`, `TRAINER_ADAPTERS`, and `TRAJECTORY_BUILDERS` for better modularity. Adjusted related models, workflow logic, and configurations to align with the new structure. This improves clarity and supports more granular component handling.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Add support for fine-tuning components in workflow builder

Introduced Trainer, TrainerAdapter, and TrajectoryBuilder components with configurations, mocking methods, and dependency tracking. Enhanced testing and registration workflows to validate the integration of these new components into the builder and workflow lifecycle.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Add reward config and comprehensive tests for finetuning modules.

This commit introduces a reward field to critical configuration models, including `TrainerConfig` and `TrajectoryBuilderConfig`, enabling enhanced reward-based customization. Additionally, it implements extensive unit tests for `Trainer`, `TrajectoryBuilder`, and `TrainerAdapter`, covering initialization, run workflows, and error handling for robust validation.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Add unit tests for NAT finetuning parsers

Introduced comprehensive unit tests for `llama_index_parser`, `langchain_parser`, and `base_parser` under `tests/nat/finetuning/utils/parsers`. These tests cover various edge cases, ensuring proper functionality of content extraction, message parsing, and sequence validation. Added `__init__.py` files to relevant directories for module structure.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Merge remote-tracking branch 'upstream/develop' into finetuning-harness

dnandakumar-nv committed:
Pre-commit checks

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Add `finetune` CLI command for workflow finetuning

Introduced a new `finetune` command to the CLI, allowing users to run finetuning on workflows using collected trajectories. This includes options for specifying a configuration file and overriding config values. Additionally, the command integrates with existing CLI infrastructure and logs progress during execution.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
`Refactor finetuning configuration handling`

Introduced a unified `FinetuneConfig` model to replace `FinetuningConfig` and `FinetuneRunConfig`, simplifying finetuning-related data handling. Updated runtime, interfaces, and CLI to leverage the new configuration model, including enhanced validation and improved CLI parameter handling.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Add num_epochs to Trainer and enhance finetuning workflow

Introduced a `num_epochs` field to the `FinetuneConfig` model for managing training iterations. Updated `Trainer` and `TrainerAdapter` to simplify initialization, validation, and cleanup processes. Refactored logic for managing configuration and validation datasets, streamlining the finetuning execution.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Refactor imports and clean up redundant code across finetuning modules

Reorganized imports to adhere to consistent style, prioritizing standard and external libraries first, followed by internal ones. Removed unused imports and redundant whitespace for better readability. Simplified code structure and formatting across multiple files to enhance maintainability and clarity.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Refactor imports and clean up redundant code across finetuning modules

Reorganized imports to adhere to consistent style, prioritizing standard and external libraries first, followed by internal ones. Removed unused imports and redundant whitespace for better readability. Simplified code structure and formatting across multiple files to enhance maintainability and clarity.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Add OpenPipe ART integration subpackage to NeMo toolkit

This commit introduces a new subpackage, `nvidia-nat-openpipe-art`, for OpenPipe ART integration in the NVIDIA NeMo Agent Toolkit. It includes configuration files such as `pyproject.toml`, licensing information, and placeholder plugin files for future expansion.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Add support for nvidia-nat-openpipe-art package

Integrates the nvidia-nat-openpipe-art package into the project, including its dependency on openpipe-art version 0.5.1. Updates uv.lock and pyproject.toml to reflect the new package and its associated configurations.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Add ART trajectory builder implementation for OpenPipe backend

This commit introduces the ARTTrajectoryBuilder and its configuration, enabling trajectory generation for the OpenPipe backend. It includes registration logic, main functionalities, and associated dependency updates.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Add ARTTrainerAdapter and enhance OpenPipe ART framework

Introduced ARTTrainerAdapter for integrating with the ART backend, enabling advanced training workflows. Updated configuration classes, registration logic, and initialization to support new functionality, including reward functions in both trajectory builder and trainer adapter.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Add ARTTrainer implementation and curriculum learning support

Introduced ARTTrainer for OpenPipe ART backend, enabling finetuning with curriculum learning and trajectory-based rewards. Updated configuration, registration, and dependencies to support these functionalities, including matplotlib for visualizations.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Update default trainer run name in configuration

Changed the default value of the `name` field in the configuration from "trainer_run_4" to "trainer_run". This simplifies the default naming convention for trainer runs.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Add tests for ARTTrainer and ARTTrainerAdapter functionality

Introduced comprehensive unit tests for `ARTTrainer` and `ARTTrainerAdapter` to validate initialization, training, trajectory handling, status management, and error handling. These tests ensure robust functionality and resilience of training components.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Add RL with OpenPipe ART Tic-Tac-Toe workflow and components

Introduces a Tic-Tac-Toe game workflow utilizing LangChain for LLM-based agents. Includes game logic, XML move parsing, heuristic evaluation, and a configurable function for LLM interaction. Adds dependency management, configuration, and registration setup for OpenPipe ART integration.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Revert uv.lock file to revision 1 and clean up metadata.

Rolled back `uv.lock` file from revision 3 to revision 1. Removed redundant `upload-time` metadata for sdist and wheel dependencies to simplify the file structure and ensure consistency.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Add missing import for workflow function registration

This ensures that the `rl_with_openpipe_art_function` is properly imported, triggering its registration as expected. This change improves the reliability of the registration process.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Add target_model filter to trajectory processing

Introduced a `target_model` attribute to allow filtering of intermediate steps by model name during trajectory processing. Updated the trajectory builder logic to skip steps not matching the specified target model, ensuring more precise fine-tuning when multiple models are involved.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Add ruff ignore flags and placeholder data file

Added `noqa` flags to suppress unused import warnings for ruff and flake8. Introduced a new placeholder data file, `data.json`, with sample question-answer pairs for testing purposes.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Add custom accuracy evaluator for RL workflow outputs

Introduce `AccuracyEvaluator` class to evaluate RL outputs based on specific scoring logic. Register the evaluator for use in workflows and include its configuration. Also, update logging in the Tic-Tac-Toe workflow from `info` to `debug` for reduced verbosity.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Add randomized move option and logging improvements

Introduced a `choose_random` flag for LLM-based Tic Tac Toe players, enabling randomized moves when specified. Enhanced logging by replacing print statements with logger calls and adjusted configuration defaults for better evaluation control.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Add finetuning configuration and update annotation handling

Introduced a comprehensive finetuning configuration to enable specific training setups, including trajectory builders, trainer adapters, and reward functions. Updated `config.py` to include new annotations and rebuild handling for trainers, trainer adapters, and trajectory builders. These changes facilitate flexible and scalable finetuning workflows.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Add support for new trainer-related keys and handle TypeError

Extend the config validator to support trainers, trainer adapters, and trajectory builders in the GlobalTypeRegistry. Additionally, update the component discovery utility to handle TypeError for types that don’t support isinstance checks, ensuring robust handling of complex type unions.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Enable fine-tuning and update required configuration fields.

Fine-tuning is now activated by default in the configuration file. Removed "config_file" as a required field for validation and updated trainer_adapter references for consistency. These changes streamline the setup process for fine-tuning workflows.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Refactor payload handling and clean up unused code.

Introduce `payload` in `IntermediateStep` for raw generation data. Remove unused methods, redundant error handling, and unused imports, simplifying callback handler logic for better maintainability.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Update temperature settings, add component registrations, and fix tests

Adjusted temperature values in the RL configuration for tuning. Introduced registration for finetuning components and added relevant imports in test files. Enhanced the `dask_cluster_fixture` for thread-based workers, improving test reliability and performance.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Refactor: Remove heuristic evaluation logic and update configs

Removed the `evaluate_board_for_player` function and related references to streamline the core logic of the RL module. Updated model configurations in `config.yml` to use new model versions for better alignment with the latest LLM capabilities. These changes simplify the codebase and enhance maintainability.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Refactor RL configs into modular files for clarity.

Split the monolithic RL fine-tuning config into three modular files: baseline, training, and post-training. This improves organization, readability, and distinction between different phases of the workflow.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Add heuristic board evaluation and cleanup config files

Introduced an Alpha-Beta-style heuristic evaluation for Tic Tac Toe boards, improving AI decision-making with detailed scoring metrics. Removed outdated configuration YAML files related to fine-tuning and replaced them with cleaner structures. This refactor improves code clarity, reduces redundancy, and enhances AI gameplay logic.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Add configs and refine RL workflow for OpenPipe ART

Introduce pre-train, finetune, and post-train configuration files for RL with OpenPipe ART, enabling LLM-specific workflows and evaluations. Refine code by removing heuristic evaluations and streamlining move evaluation logic for improved clarity and maintainability.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Simplify config structures and update default parameters.

Replaced default_factory with None for clarity in backend configs. Adjusted file paths and updated temperature values, model names, and various training parameters for consistency and better evaluation alignment. Removed redundant settings to streamline configuration files.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Refactor player logic and evaluation in RL workflow.

Updated the RL workflow to handle player model configurations more flexibly, simplified move retries, and refined game termination handling. Added failure penalty in evaluation, adjusted default learning rates, and updated dataset answers to reflect new scoring outputs.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Simplify message tracking and improve evaluation logic

Modified parsing logic to prevent duplicate messages, improved evaluation logic for scoring aborted games, and adjusted configurations for training, evaluation, and model setup. Additional enhancements include refining message history handling and making the board display more user-friendly.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Update trainer logic and configuration defaults

Adjusted "finish_reason" in trainer adapter to "stop" for consistency and updated training beta default to 0 for improved behavior. Additionally, set max_parser_retries default to 0 in RL example to align with updated retry strategy.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Refactor configurations and logic for improved evaluation.

Updated configuration files to adjust temperature, concurrency, and dataset paths. Refined game logic to account for player roles. Simplified trajectory message handling and expanded evaluation dataset with additional examples.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Refactor RL evaluator and game logic for improved traceability

Updated the RL evaluator to incorporate intermediate step value computation for richer insights during evaluation. Refactored game logic to enable step-level state tracking with value assignments, improving training feedback. Adjusted configurations for model, evaluator, and workflow parameters to align with the updated logic.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Adjust training configs and expand dataset for RL module

Updated training parameters, including learning rate, temperature, and number of epochs. Adjusted naming conventions for clarity and added `max_parser_retries` to configs. Expanded the dataset with additional entries to improve model training coverage.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Add finetuning documentation and OpenPipe ART integration

This update introduces documentation on the finetuning harness, covering key concepts, reinforcement learning approaches, and curriculum learning. Additionally, it includes a reference for integrating OpenPipe ART for training LLMs using GRPO optimization. Configuration and CLI usage details are also provided.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Add RL Tic-Tac-Toe example using OpenPipe ART

This commit introduces a detailed example demonstrating reinforcement learning finetuning with OpenPipe ART to enhance an LLM's Tic-Tac-Toe performance. Includes a comprehensive README explaining setup, workflow, reward function, and best practices, alongside necessary configs, scripts, and assets.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Merge remote-tracking branch 'upstream/develop' into finetuning-harness

dnandakumar-nv committed:
Update comment for dependency overrides in pyproject.toml

Clarified the purpose of overriding dependencies to prevent regressions caused by nvidia-nat-openpipe-art. Added additional context to emphasize maintaining the status quo for other packages.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Fix relative link paths to Custom Evaluators in docs

Updated the relative paths to the Custom Evaluators page to ensure proper navigation in the documentation. This resolves incorrect linking issues in the finetuning reference guides.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Fix typo in 'logprob' vocabulary entry in Vale config

Corrected the case-insensitivity pattern for 'logprob' to ensure proper matching in the Vale style vocabulary. This update aligns with consistent handling of vocabulary terms.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Refactor module naming and improve setup instructions

Updated module and project naming to align with the "nat" prefix convention. Adjusted logging suppression prefix and enhanced setup instructions in README for better clarity, including specifying virtual environment and dependencies.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Refactor module naming and improve setup instructions

Updated module and project naming to align with the "nat" prefix convention. Adjusted logging suppression prefix and enhanced setup instructions in README for better clarity, including specifying virtual environment and dependencies.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Update README for ART setup with corrected Python version and package

Changed the Python version to 3.11 and specified the exact OpenPipe ART package version for consistent setup. This ensures compatibility and avoids potential installation issues.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Add DPO Tic-Tac-Toe example with utilities and workflow

Introduce a detailed example of a Tic-Tac-Toe workflow for DPO preference data generation using Test Time Compute. Includes core logic for board evaluation, LangChain-based LLM agents, evaluators for DPO training, and configuration files.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Refactor Tic-Tac-Toe workflow to integrate TTC move selector

Replaced individual TTC components (search, score, and select) with the consolidated `ttc_move_selector` function. This simplifies the workflow by delegating the full move selection pipeline, including candidate generation and recording intermediate steps, to the selector function. Added support for the `multi_candidate_move_search` strategy.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Add random move generation and TTC support for all turns

Integrated random move generation for non-LLM opponents and enabled the TTC pipeline for both trained players and opponents. Updated configuration and workflows to support DPO data collection from all game turns, improving consistency and flexibility in model training and evaluation.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Add prompt handling to enhance DPO and move selection logic

Introduced `format_prompt_for_dpo` and `get_system_prompt` to support structured prompt generation, improving data flow for DPO training. Updated move-related functions to include prompts in outputs, ensuring consistency and capturing full context for debugging and downstream processes.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Refactor prompt handling for choose_move function

Simplify prompt management by using the board state string directly instead of constructing or formatting prompts. Removes unused functions and streamlines logic in random and LLM modes for consistency.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Add DPO Trajectory Builder and configuration for NAT finetuning

Introduces the DPO (Direct Preference Optimization) Trajectory Builder, enabling the generation of preference data for finetuning workflows in NAT. Added related configuration, plugins, and registration support, as well as setup changes for package integration. This enhancement supports custom trajectory building logic for DPO training.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Reduce redundant line breaks and enhance code formatting

Removed unnecessary blank lines and improved inline formatting for cleaner and more consistent code structure. This simplifies readability across multiple files without altering functionality.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Add tests for the NeMo Customizer NAT plugin

This commit introduces comprehensive test coverage for the NeMo Customizer NAT plugin, focusing on the DPO Trajectory Builder. The tests validate configuration behaviors, candidate parsing, preference pair generation, trajectory building, and various edge cases to ensure robust functionality and correctness.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Add DPO pipeline configuration and documentation for Tic-Tac-Toe

Introduced a detailed DPO finetuning pipeline and enhanced configuration for preference data generation. Added a comprehensive README explaining the workflow, architecture, key components, and usage instructions.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Enhance token usage tracking and tool output handling.

Refactor token usage extraction to capture additional details like cached and reasoning tokens. Add robust handling of tool outputs in callbacks, improving support for server-side tool usage and error resilience during schema parsing.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Add finetune command documentation to CLI reference

This update introduces the `nat finetune` command, providing detailed documentation for its usage, options, and examples. The finetune command allows in-situ reinforcement learning for iterative agent improvement, incorporating experimental features and extensive configuration options.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Add support for NeMo Customizer plugin and TTC event handling

Integrated the NeMo Customizer plugin into the project and updated dependencies accordingly. Introduced Test-Time Computation (TTC) event handling in intermediate step models and modified related functions for consistency. Enhanced DPO finetuning workflows with support for OpenAI message formatting and improved data structures for conversation handling.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Delete unused `spec` module in `api_utils`.

The `spec` module is no longer in use and has been removed to streamline the codebase. This cleanup helps reduce clutter and improves maintainability.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Add score field and support DPOItem in episode structure.

Introduced a new `score` field in `IntermediateStep` for enhanced granularity. Updated trajectory episodes to support both `EpisodeItem` and `DPOItem`, while refactoring `score` handling in `ttc_move_selector_function` for consistency.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Refactor TTC data handling and migration to structured model

This commit introduces a transition from flexible metadata-based handling (dicts) to using the structured TTCEventData model for test-time preferences. The new implementation improves type safety and robustness by directly extracting fields like turn_id, candidate_index, score, input, and output from TTCEventData. It removes the need for configurable dictionary key mappings, simplifying workflows and enhancing maintainability.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Add NeMo Customizer TrainerAdapter with DPO support

Introduced the `NeMoCustomizerTrainerAdapter` for submitting DPO/SFT training jobs to NeMo Customizer, including optional deployment configuration. Enhanced the DPO module with new components, registration updates, and extensive test coverage for hyperparameter configurations, deployment settings, and adapter functionality.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Add NeMo Customizer Trainer for DPO finetuning with tests

Introduce the NeMo Customizer Trainer for handling DPO data collection and training. This includes configuration support, trajectory handling, training job orchestration, and necessary integration updates. Also added comprehensive unit tests for config validation and trainer behavior.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Update endpoint URLs in config and register files

Replaced outdated NVIDIA-specific endpoint URLs with generic placeholder URLs (`example.com`) to improve configurability and remove hardcoded dependencies. This ensures the code is more adaptable to different environments.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Delete obsolete NeMo Customizer and test modules

Remove outdated NeMo Customizer API utilities, test configuration files, and test cases from the repository. The deletion cleans up unused code and ensures the repository contains only relevant and current resources.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Set default to exclude full message history and enhance logging.

Changed `use_full_message_history` default to `False` to simplify prompt formatting. Added detailed logging in `trainer_adapter` to track job status and progress updates more clearly.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Refactor serialization, config, and validation logic.

Updated `IntermediateStep` to use `SerializeAsAny` for the `data` field for better type flexibility. Enhanced DPO config with new fields for preference and SFT loss. Simplified health check in `trainer_adapter` and adjusted fine-tuning validation by removing `reward_function` from required fields.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Update config and trainer for NeMo DPO fine-tuning

Upgraded model to `meta/llama-3.3-70b-instruct` and adjusted concurrency for evaluation. Added NeMo Customizer trainer configurations, enabling fine-tuning and deployment for the DPO workflow. Simplified dataset repository naming in the trainer adapter.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Update default dataset name in configuration

Changed the default value of `dataset_name` from "nat-dpo-dataset" to "nat-dpo" for consistency and clarity. This ensures better alignment with naming conventions.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Update deployment handling and logging for DPO training

Switched training LLM configuration to OpenAI-compatible settings and added a deployment timeout parameter. Improved model deployment process by adding a readiness check with polling and timeout. Simplified and clarified logging for deployment and training progress.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Remove unused code and enhance tic-tac-toe README.

Deleted outdated `scartch.py` file for cleaner project structure. Updated tic-tac-toe README with detailed installation guides, configuration references, and expanded pipeline architecture for preference learning with NeMo Customizer.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Merge remote-tracking branch 'upstream/develop' into finetuning-harness

# Conflicts:
#	uv.lock

dnandakumar-nv committed:
Add support for `openpipe-art` package and related dependencies

Integrated `openpipe-art` package and its dependencies, including `litellm` and `openai`. Updated project metadata, extras, and lockfile entries to enable its functionality. This enhancement introduces new examples and configurations for fine-tuning workflows.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Update OpenPipe ART integration and dependencies

Updated the description for the OpenPipe ART subpackage in `pyproject.toml` to reflect proper integration context. Added `nvidia-nat-openpipe-art` as a dependency in the `nvidia_nat_all` package for consistency in workspace management.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Add guideline to update dependencies in `nvidia-nat-all`

Clarifies that when a new package is added, its name must be included as a dependency in the `nvidia-nat-all` package's `pyproject.toml` file. Ensures consistency in dependency management.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Add ADK framework support and shared parser utilities

Introduced ADK-specific parser handling logic for LLM messages and integrated it into the existing framework switch. Refactored common parsing utilities for reusability, reducing code duplication in LangChain and LlamaIndex parsers. Added comprehensive unit tests for ADK parser functionality, ensuring proper content extraction and message parsing across edge cases.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Refine formatting, comments, and docstrings for consistency.

Updated bullet point styles in comments for uniformity across files. Standardized docstring formatting, ensuring proper punctuation and clear argument descriptions. These changes improve readability and maintain documentation consistency.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Merge branch 'finetuning-harness' into finetuning-harness-dpo

# Conflicts:
#	uv.lock

dnandakumar-nv committed:
Add A2A packages and update DPO configuration references

Added new A2A-related packages: `nvidia-nat-a2a`, `nat-currency-agent-a2a`, and `nat-math-assistant-a2a`. Adjusted DPO training configurations to use dynamic environment variables for NeMo Customizer endpoints and streamlined example documentation. Removed redundant LoRA vs full-weight finetuning section from docs for clarity.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Merge remote-tracking branch 'upstream/develop' into finetuning-harness

dnandakumar-nv committed:
Add post-training config and evaluation steps for DPO Tic-Tac-Toe

Introduce a new configuration file for post-training workflows and document the evaluation process in the README. The updates enable model evaluation using a deployed LLM with NAT pipelines and provide detailed steps for reproducing results.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Merge branch 'finetuning-harness' into finetuning-harness-dpo

dnandakumar-nv committed:
Add nvidia-nat-nemo-customizer to dependencies and workspace

This commit includes the nvidia-nat-nemo-customizer package in both the dependency list and the workspace settings in pyproject.toml. This ensures the package is properly tracked and built as part of the project.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dagardner-nv commented on examples/finetuning/dpo_tic_tac_toe/pyproject.toml:
```diff
@@ -0,0 +1,26 @@
+dynamic = ["version"]
+dependencies = [
+  "nvidia-nat[langchain]~=1.4",
+  "numpy>=1.24.0"
```
This is a different version of numpy than the one specified in the top-level pyproject.toml file, update to match.
```suggestion
  # langchain-aws pins numpy to 1.* for Python <3.12 -- remove once package is updated
  "numpy~=1.26; python_version < '3.12'",
  "numpy~=2.3; python_version >= '3.12'",
```

dagardner-nv commented on packages/nvidia_nat_nemo_customizer/pyproject.toml:
```diff
@@ -0,0 +1,55 @@
+  # version when adding a new package. If unsure, default to using `~=` instead of `==`. Does not apply to nvidia-nat packages.
+  # Keep sorted!!!
+  "nvidia-nat~=1.4",
+  "nemo-microservices==1.4.0"
```
Any reason to use == instead of ~=?
```suggestion
  "nemo-microservices~=1.4"
```

dagardner-nv commented on pyproject.toml:
```diff
@@ -213,6 +222,7 @@ nat_plot_charts = { path = "examples/custom_functions/plot_charts", editable = t
 nat_por_to_jiratickets = { path = "examples/HITL/por_to_jiratickets", editable = true }
 nat_profiler_agent = { path = "examples/advanced_agents/profiler_agent", editable = true }
 nat_router_agent = { path = "examples/control_flow/router_agent", editable = true }
+nat_rl_with_openpipe_art = { path = "examples/finetuning/rl_with_openpipe_art", editable = true }
```
Don't forget to add the tic tac toe example

dnandakumar-nv committed:
Add licensing files for NVIDIA package dependencies

This commit introduces the Apache License 2.0 and third-party license files to the NVIDIA package. These files provide compliance with open-source licensing requirements by specifying the terms and conditions of use, reproduction, and distribution for the package and its dependencies.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Remove outdated test README and add missing SPDX headers

Deleted the `tests/README.md` file as it was outdated and redundant. Added SPDX license headers to multiple config files and updated file paths in configuration files for improved clarity. Also included new terms in the vocabulary acceptance list.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Add new path patterns for openpipe_art to path checks

Updated the path checks script to include rules for

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Fix typo in vocabulary list for "finetuning" and "finetunable"

Corrected capitalization issues in the listed terms to align with consistent styling. This ensures better accuracy and functionality in text validation tools.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Refine path checks and update vocabulary/guide.

Improved regex patterns in path checks for better accuracy and coverage. Expanded accepted vocabulary in Vale configurations with finetuning-related terms. Added path-check skip markers in the finetuning documentation for clarity and alignment with new rules.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Update vocab and refine headers in RL fine-tuning docs

Adjusted vocabulary to include "subsampling" and fix case typo for "rollout". Renamed headers for clarity in documentation, providing consistent class references for `ARTTrainerAdapter` and `ARTTrainer`.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Expand path checks with additional config patterns

Updated the path checks to include a specific config.yaml file path under the `src/rl_with_openpipe_art/configs` directory. This ensures proper handling and validation of the newly added configuration structure.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Add 'Qwen/Qwen.*' pattern to path checks

This update includes the 'Qwen/Qwen.*' regex pattern in the path checks to expand the validation ruleset. It ensures proper handling of paths related to the newly added category.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Merge branch 'finetuning-harness' into finetuning-harness-dpo

dnandakumar-nv committed:
Merge remote-tracking branch 'upstream/develop' into finetuning-harness-dpo

# Conflicts:
#	packages/nvidia_nat_all/pyproject.toml
#	pyproject.toml
#	src/nat/data_models/finetuning.py
#	src/nat/data_models/intermediate_step.py
#	uv.lock

dnandakumar-nv committed:
Update packages/nvidia_nat_nemo_customizer/pyproject.toml

Co-authored-by: David Gardner <96306125+dagardner-nv@users.noreply.github.com>
Signed-off-by: Dhruv Nandakumar <168006707+dnandakumar-nv@users.noreply.github.com>

dnandakumar-nv committed:
Add `nvidia-nat-nemo-customizer` and `nat-dpo-tic-tac-toe`

Introduced the `nvidia-nat-nemo-customizer` package with dependencies and made `nat-dpo-tic-tac-toe` editable within examples. Updated related metadata, dependencies, and extras in the lock and TOML files accordingly.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Merge remote-tracking branch 'origin/finetuning-harness-dpo' into finetuning-harness-dpo

dnandakumar-nv committed:
Refactor DPO Tic-Tac-Toe structure and update licensing.

Relocated DPO Tic-Tac-Toe data, configs, and README files into a src directory for better organization. Added licensing details and "path-check-skip" tags where appropriate. Included third-party license files for legal compliance.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Update examples/finetuning/dpo_tic_tac_toe/pyproject.toml

Co-authored-by: David Gardner <96306125+dagardner-nv@users.noreply.github.com>
Signed-off-by: Dhruv Nandakumar <168006707+dnandakumar-nv@users.noreply.github.com>

dnandakumar-nv commented:
@coderabbitai review

dnandakumar-nv commented:
/ok to test 93315c5b6bf42254f8ea3ffd136151687bbf91b0

dnandakumar-nv committed:
Use consistent YAML syntax, fix copyright years, and clarify docs

Standardize YAML block formatting throughout code and docs for better readability. Update SPDX copyright years and expand license headers. Fix minor terminology inconsistencies in documentation and improve clarity in explanations.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Add link to NeMo Customizer documentation in finetuning guide

This update includes a reference to the NeMo Customizer in the finetuning guide's index. It ensures better accessibility to related documentation for users exploring DPO workflows.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv commented:
/ok to test bab8b57860b5384e9a8ecb4bd752a08fc59a80fc

dnandakumar-nv committed:
Refactor finetuning docs: remove in-situ training section.

Removed redundant in-situ training explanation from the finetuning concepts documentation and improved clarity on DPO training with new code snippets. Also adjusted minor formatting inconsistencies in the NeMo Customizer guide.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv commented:
/ok to test a5903a52e28cb4a2726555b521c06f3d8e600121

dnandakumar-nv committed:
Update regex for DPO Tic Tac Toe data path validation

Adjusted the regex pattern to ensure it matches optional subdirectories in the DPO Tic Tac Toe data path. This enhances the flexibility of the path validation logic for more robust use cases.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv commented:
/ok to test 228e5eaac5cbf1d0fd26f65a39e57cf0e885c513

dnandakumar-nv committed:
Update default config values and simplify health check test

Updated default deployment configurations to reflect new image and message history settings. Simplified the health check test by switching to a stub implementation that always returns True, removing unnecessary mocking logic.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

willkill07 commented on src/nat/data_models/finetuning.py:
```diff
@@ -123,11 +123,28 @@ def check_logprobs(self) -> "EpisodeItem":
         return self
 
 
+class OpenAIMessage(BaseModel):
```
Sometimes there are extra components to OpenAIMessage like tool call information. Is that important here?

dnandakumar-nv replied:
It shouldn't be. DPO and most other backends only need this much information. I think we can also extend this later with fields as needed.

willkill07 commented on packages/nvidia_nat_nemo_customizer/src/nat/plugins/customizer/dpo/__init__.py:
```diff
@@ -0,0 +1,44 @@
+__all__ = [
+    # Trajectory Builder
+    "DPOTrajectoryBuilderConfig",
+    "DPOTrajectoryBuilder",  # Trainer
+    "NeMoCustomizerTrainerConfig",
+    "NeMoCustomizerTrainer",  # TrainerAdapter
+    "NeMoCustomizerTrainerAdapterConfig",
```
If you want the formatter to work on behalf of you:
```suggestion
    "DPOTrajectoryBuilder",  #
    # Trainer
    "NeMoCustomizerTrainerConfig",
    "NeMoCustomizerTrainer",  #
    # TrainerAdapter
    "NeMoCustomizerTrainerAdapterConfig",
```

willkill07 commented on packages/nvidia_nat_nemo_customizer/src/nat/plugins/customizer/__init__.py:
```diff
@@ -0,0 +1,43 @@
+__all__ = [
+    # Trajectory Builder
+    "DPOTrajectoryBuilder",
+    "DPOTrajectoryBuilderConfig",  # TrainerAdapter
+    "NeMoCustomizerTrainerAdapter",
```
```suggestion
    "DPOTrajectoryBuilderConfig",  #
    # TrainerAdapter
    "NeMoCustomizerTrainerAdapter",
```

willkill07 reviewed (approved):
Approving with minor feedback

dnandakumar-nv committed:
Update packages/nvidia_nat_nemo_customizer/src/nat/plugins/customizer/dpo/__init__.py

Co-authored-by: Will Killian <2007799+willkill07@users.noreply.github.com>
Signed-off-by: Dhruv Nandakumar <168006707+dnandakumar-nv@users.noreply.github.com>

dnandakumar-nv committed:
Update packages/nvidia_nat_nemo_customizer/src/nat/plugins/customizer/__init__.py

Co-authored-by: Will Killian <2007799+willkill07@users.noreply.github.com>
Signed-off-by: Dhruv Nandakumar <168006707+dnandakumar-nv@users.noreply.github.com>

dnandakumar-nv committed:
Refactor comments in `__all__` for clarity and consistency.

Updated inline comments in `__all__` lists to ensure accurate references to associated components (e.g., Trainer, TrainerAdapter). This improves readability and avoids misleading annotations.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv commented:
/ok to test 001515678a11ee72b3aa7ec4a3d54d5bdfe0564c

dnandakumar-nv committed:
Remove references to deprecated NeMo Customizer guide.

The links and references to the NeMo Customizer guide were deleted as the guide is no longer relevant. This change ensures that documentation remains accurate and up to date.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv committed:
Merge remote-tracking branch 'upstream/develop' into finetuning-harness-dpo

# Conflicts:
#	docs/source/improve-workflows/finetuning/dpo_with_nemo_customizer.md

dnandakumar-nv committed:
Add DPO with NeMo Customizer documentation.

Introduced comprehensive documentation for Direct Preference Optimization (DPO) training with NeMo Customizer. Includes installation steps, configuration examples, workflow implementation, and training execution instructions.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv commented:
/ok to test df4e75302e86e7f0827c7b0d988df736294002c5

dnandakumar-nv committed:
Update broken and outdated links in DPO finetuning docs

Corrected relative paths for internal documentation links to ensure they point to the correct locations. This improves navigability and prevents confusion for users exploring related resources.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv commented:
/ok to test 111ee0238af93c3b4b71ca69cbd99c661d5ccc6b

dnandakumar-nv committed:
Fix documentation formatting and update outdated content.

Adjusted diagram formatting for consistency and removed redundant steps related to model usage. Fixed minor grammatical issues and updated broken or outdated internal links for accuracy.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv commented:
/ok to test c2639ec1d11bce414f60a77deec3cfdf1be620dc

dnandakumar-nv committed:
Update contributing link in PR template

Corrected the link to the Contributing Guidelines to point to the accurate location. This ensures contributors are directed to the correct resource for guidelines.

Signed-off-by: dnandakumar-nv <dnandakumar@nvidia.com>

dnandakumar-nv commented:
/ok to test 1f15c070897120e67651c8b0efa94cae53ec3434

dnandakumar-nv commented:
/merge
